daemon/api.go

```diff
@@ -1581,6 +1582,73 @@ var (
 	osutilAddUser                = osutil.AddUser
 )
 
+func getUserDetailsFromStore(email string) (string, *osutil.AddUserOptions, error) {
+	v, err := storeUserInfo(email)
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot create user %q: %s", email, err)
+	}
+	if len(v.SSHKeys) == 0 {
+		return "", nil, fmt.Errorf("cannot create user for %s: no ssh keys found", email)
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, v.OpenIDIdentifier)
+	opts := &osutil.AddUserOptions{
+		SSHKeys: v.SSHKeys,
+		Gecos:   gecos,
+	}
+	return v.Username, opts, nil
+}
+
+func getUserDetailsFromAssertion(st *state.State, email string) (string, *osutil.AddUserOptions, error) {
+	st.Lock()
+	db := assertstate.DB(st)
+	modelAs, err := devicestate.Model(st)
+	st.Unlock()
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot get model assertion: %s", err)
+	}
+
+	brandID := modelAs.BrandID()
+	series := modelAs.Series()
+	model := modelAs.Model()
+
+	a, err := db.Find(asserts.SystemUserType, map[string]string{
+		"brand-id": brandID,
+		"email":    email,
+	})
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot get system-user assertion: %s", err)
+	}
+	su, ok := a.(*asserts.SystemUser)
+	if !ok {
+		return "", nil, fmt.Errorf("Invalid assert %s", a)
+	}
+
+	// cross check that the assertion is valid for the given series/model
+	contains := func(needle string, haystack []string) bool {
+		for _, s := range haystack {
+			if needle == s {
+				return true
+			}
+		}
+		return false
+	}
+	if !contains(series, su.Series()) {
+		return "", nil, fmt.Errorf("%q not in series %q", series, su.Series())
+	}
+	if !contains(model, su.Models()) {
+		return "", nil, fmt.Errorf("%q not in models %q", model, su.Models())
+	}
+
```

Comment:
Also, we need to enforce the until date on signature. I suggest starting on the conservative side, and refusing to sign such assertions if they're valid for more than 367 days. Will raise that on the previous PR.